plumbing/format/revfile/decoder.go

```diff
@@ -0,0 +1,226 @@
+package revfile
+
+import (
+	"bufio"
+	"bytes"
+	"crypto"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"hash"
+	"io"
+	"sync"
+
+	"github.com/go-git/go-git/v6/plumbing"
+	"github.com/go-git/go-git/v6/utils/binary"
+)
+
+var (
+	// ErrUnsupportedVersion is returned by Decode when the rev file version
+	// is not supported.
+	ErrUnsupportedVersion = errors.New("unsupported version")
+	// ErrMalformedRevFile is returned by Decode when the rev file is corrupted.
+	ErrMalformedRevFile = errors.New("malformed rev file")
+	// ErrUnsupportedHashFunction is returned by Decode when the rev file defines an
+	// unsupported hash function.
+	ErrUnsupportedHashFunction = errors.New("unsupported hash function")
+	// ErrEmptyReverseIndex is returned by Decode when the rev file is empty.
+	ErrEmptyReverseIndex = errors.New("reverse index is empty")
+
+	revHeader = []byte{'R', 'I', 'D', 'X'}
+)
+
+const (
+	VersionSupported        = 1
+	sha1Hash         uint32 = 1
+	sha256Hash       uint32 = 2
+)
+
+// Decoder reads and decodes idx files from an input stream.
+type Decoder struct {
+	reader  *bufio.Reader
+	hasher  crypto.Hash
+	hash    hash.Hash
+	nextFn  stateFn
+	version uint32
+
+	objCount     int64
+	packChecksum plumbing.ObjectID
+	out          chan<- uint32
+
+	m sync.Mutex
```

Comment:
I understand the intention of using mutex to protect against misuse but it complicates and it's inherently single use.

How about having standalone `Decode` function and unexporting `Decoder` such that each `Decode` call would initialize a new decoder where there is no need for mutex, and free of mutation bugs?

Happy to make the change if you agree